pkg/cloudprovider/gcp/object_store.go

```diff
@@ -76,16 +91,24 @@ func (o *objectStore) Init(config map[string]string) error {
 	}
 	o.client = client
 
+	o.bucketWriter = &writer{client: o.client}
+
 	return nil
 }
 
 func (o *objectStore) PutObject(bucket string, key string, body io.Reader) error {
-	w := o.client.Bucket(bucket).Object(key).NewWriter(context.Background())
-	defer w.Close()
+	w := o.bucketWriter.getWriteCloser(bucket, key)
 
+	// The writer returned by NewWriter is asynchronous, so errors aren't guaranteed
+	// until Close() is called
 	_, err := io.Copy(w, body)
 
-	return errors.WithStack(err)
+	// Ensure we close w and report errors properly
+	if closeErr := w.Close(); err == nil {
```

Comment:
This is confusing to me - normally you check the LHS error as the RHS condition. Would you mind splitting this up so it's more like this
```go
closeErr := w.Close()
if err != nil { return err }
return closeErr
```

probably should rename `err` to `copyErr`